flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/action/ActionBase.java

```diff
@@ -61,4 +93,48 @@ public abstract class ActionBase implements Action {
             throw new RuntimeException(e);
         }
     }
+
+    /**
+     * Extract {@link LogicalType}s from Flink {@link org.apache.flink.table.types.DataType}s and
+     * convert to Table Store {@link DataType}s.
+     */
+    protected List<DataType> toTableStoreDataTypes(
+            List<org.apache.flink.table.types.DataType> flinkDataTypes) {
+        return flinkDataTypes.stream()
+                .map(org.apache.flink.table.types.DataType::getLogicalType)
+                .map(LogicalTypeConversion::toDataType)
+                .collect(Collectors.toList());
+    }
+
+    /** Get {@link DataType}s from table's schema. */
+    protected List<DataType> getDataType() {
+        return ((FileStoreTable) table).schema().logicalRowType().getFieldTypes();
+    }
+
+    /**
+     * Check whether each {@link DataType} of actualTypes can be implicitly cast to that of
+     * expectedTypes respectively.
+     */
+    protected boolean checkSchema(List<DataType> actualTypes, List<DataType> expectedTypes) {
+        if (actualTypes.size() != expectedTypes.size()) {
+            return false;
+        }
+
+        for (int i = 0; i < actualTypes.size(); i++) {
+            if (!DataTypeCasts.supportsImplicitCast(actualTypes.get(i), expectedTypes.get(i))) {
```

Comment:
Int -> Long will throw cast exception.
We should introduce a rule to `DataTypeCasts`, for example, `dataStructureCastingRule`.